doc/categorise_focus.man
.TH CATEGORISE_FOCUS 1
.SH NAME
CATEGORISE_FOCUS \- Utility script to arrange images into directories
.SH SYNOPSIS
.B CATEGORISE_FOCUS
[\fB\-v\fR]
[\fB\-\-verbose\fR]
[\fB\-d\fR]
[\fB\-k\fR \fIkey\fR]
[\fB\-\-fkey\fR \fIkey\fR]
[\fB\-r\fR]
[\fB\-\-rename\fR]
best [worst]
.IR \< scorefile ...
.SH DESCRIPTION
.B CATEGORISE_FOCUS
A utility script which groups images into BEST , WORST and the remainder.
You can do this based on the output of \fBcheckfocus\fR or \fBbestfocus\fR, and on any the data including exposure.
You can recuse down into these directories and repeat the exercise with different settings, e.g. exposure within
focus (or vice versa) or centre focus within overall focus.
.SH OPTIONS
.Tp
.BR \-v ", " \-\-verbose\fR
Enables verbose mode. It can be repeated to make output more verbose.
.PP

.Tp
.BR \-k \fI\ KEY\ \fR ", "
.BR \-\-key \fI\ KEY\ \fR
KEY is a column number in the scorefile. It must be 2 or greater (as column 1 is the filename) default is 2, which is overall score from \fBcheckfocus\fR
and the box score in \fBbestfocus\fR
.PP

.Tp
.BR \-r ", " \-\-rename\fR
Causes the moved files to be renamed to add a prefix of the chosen score (from key) so IMG123.jpg might get called BEST/13045670-IMG123.jpg

.SH EXAMPLES
With
.BR CHECKFOCUS
command:
.PP
.nf
.RS

$ find . -mtime -1       > filelist
$ checkfocus -f filelist > scorefile
$ categorise_focus 7300000 100000 < scorefile > command-file
$ ...possibly edit the command-file...
$ bash ./command-file

.RE
.fi
.PP
This simply moves the images into BEST & WORST directories (or leaves them in-place)

With 
.BR BESTFOCUS
command:
.PP
.nf
.RS

$ find . -mtime -1                        > filelist
$ bestfocus   -f filelist                 > scorefile
$ categorise_focus -r 7300000 < scorefile > command-file
$ ...possibly edit the command-file...
$ bash ./command-file

.RE
.fi
.PP
Here we use \fBbestfocus\fR, only select for BEST (no WORST directory will be created)
files moved into BEST will be prefixed with their BESTBOX score
.PP
.BR "Workflow orientated"
command:
.PP
.nf
.RS
$ find . -mtime -1                        > filelist
$ checkfocus -u 0 -o 255  -f filelist     > scorefile
$ categorise_focus -k8 95     < scorefile > command-file
	...the directory called BEST will hold images with >95% overexposed pixels
$ categorise_focus -k9 95     < scorefile >> command-file   # Note the >>
	...add to BEST images with >95% underexposed pixels
$ bash ./command-file  # Do the moves
$ mv BEST BAD-EXPOSURE # so called BEST is just biggest (over & under exposure scores)
$ categorise_focus 1900000000 1000000000  < scorefile > command-file
$ bash ./command-file  # move the good & bad focus (the poor exposure have already been removed)
.RE
.fi
.PP
Here we did multiple passes. First we moved the over and under exposed images out, then we split
the remaining files into BEST & WORST (and remainder)
.PP
The command categorise_focus just emits shell commands. Once you run these
commands you are moving files around a renaming them. In general do this with
copies of you files (or hardlinks) to avoid the risk of losing a files due to
an injudicious rename.


categorise_focus.sh
# !/bin/bash -ue

# Categorise photos by focus quality
# 

PROG=$0

set -ue 

usage() {

    cat <<-EOF
	USAGE: categorise [-v][--verbose] [-tSEP] [--field-separator=SEP] [-k key][--key key][-r][--rename] best [worst] < scorefile  > command-file

	-h or --help emits this help text
	-v or --verbose does the obvious thing
	-k key or --key key selects which key (column) is used to decide best/worst defaults to 2 (whole image from checkfocus)
	-r or --rename causes the output commands to rename the image file to prefix the chosen score (key) to filename, allows easy review in order
	-t or --field-separator, needs to match the -t option of the generating tool, to cope with filenames containing spaces

	<best> and <worst> are scores. 
	Images with values above <best> are categorised as good
	Images with values below <worst> are categorised as bad

	This script generates a set of shell commands to move the good images into BEST directory and bad images into WORST directory
	normally you would redirect the output into a file. Edit it if required, then execute it:

	find . -mtime -1       > filelist
	checkfocus  -f filelist > scorefile
	${PROG} 7300000 100000 < scorefile > command-file
	...possibly edit the command-file...
	bash ./command-file
EOF
}

typeset -i verbose 
typeset -i best
typeset -i worst
typeset -i keycolumn
typeset -i keyelement
typeset -i n
typeset -i col
typeset -i lineno
typeset -i help

typeset -a line


PROG=$0
verbose=0
rename=0
keycolumn=2

help=0


: ${best:=0}
: ${worst:=0}

sep=""

TEMP=`getopt -o hvrk:t: --long verbose,rename,key:,field-separator: -n ${PROG} -- "$@"`

if [ $? != 0 ] ; then echo "Getopt failed, terminating..." >&2 ; exit 1 ; fi

# Note the quotes around `$TEMP': they are essential!
eval set -- "$TEMP"

while true ; do
	case "$1" in
		-h|--help)            help=1         ; shift   ;;
		-v|--verbose)         verbose=1      ; shift   ;;
		-r|--rename)          rename=1       ; shift   ;;
		-k|--key)             keycolumn="$2" ; shift 2 ;;
		-t|--field-separator) sep="$2"       ; shift 2 ;;
		--) shift ; break ;;
		*) echo "Internal error!" ; exit 1 ;;
	esac
done


if [[ ${help} -ne 0 ]] ; then
  usage
  exit 1
fi


if [ $# -lt 1 -o  $# -gt 2 ] ; then
  usage
  exit 1
fi

best=$1

if [ $# -gt 1 ] ; then
    worst=$2
fi


if [ ${verbose} -eq 1 ] ; then
    echo "verbose selected"             >&2
    echo "key=${keycolumn}"             >&2
    echo "best=${best}"                 >&2
    echo "worst=${0}"                   >&2
fi

if [[ ${keycolumn} -lt 2 ]] ; then
    echo "Error key must be 2 or greater (column 1 is the filename)" >&2
    exit 1
fi

keyelement=keycolumn-1

if [[ ${best} -ne 0 ]] ; then
    echo "mkdir -p BEST"
fi

if [[ ${worst} -ne 0 ]] ; then
    echo "mkdir -p WORST"
fi

lineno=0

if [[ "${sep}" != "" ]] ; then
    OLDIFS="${IFS}"
    IFS="${sep}"
fi


while read -a line
do
    lineno+=1
    
    if [[ verbose -ne 0 ]] ; then
	n=${#line[*]}
	echo -n "${n} items on line ${lineno} " >&2
	echo -n "Filename: ${line[0]} "         >&2
	echo -n "Fields: "                      >&2
	echo -n "IFS: ${IFS}"                   >&2

	for ((i=1; i<n; ++i))
	do
	    col=i+1
	    echo -n "k[${col}]=${line[i]},"    >&2  # sort is 1 based, arrrays are 0 based
	done
	echo                                   >&2
    fi
	
    if [[ ${best} -ne 0 ]] ; then
	if [[ line[keyelement] -gt best ]] ; then
	    if [[ ${rename} -eq 0 ]] ;then
		echo "mv ${line[0]} BEST"
	    else
		echo "mv ${line[0]} BEST/${line[keyelement]}-${line[0]}"
	    fi
	    
	fi
    fi

    if [[ ${worst} -ne 0 ]] ; then
	if [[ line[keyelement] -lt worst ]] ; then
	    if [[ ${rename} -eq 0 ]] ;then
		echo "mv ${line[0]} WORST"
	    else
		echo "mv ${line[0]} WORST/${line[keyelement]}-${line[0]}"
	    fi
	    
	fi
    fi
	
done

exit

